lib/wallet/test/unit/Cardano/Wallet/Submissions/Generation.hs

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE NoMonoLocalBinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Cardano.Wallet.Submissions.Generation
+    ( prop_submission_history
+    , SubmissionsHG (..)
+    , txG
+    , slotG
+    , noSubmissions
+    , Step (..)
+    , P
+    )
+where
+
+import Prelude
+
+import Cardano.Wallet.Submissions.Properties.Common
+    ( Step (..), newState )
+import Cardano.Wallet.Submissions.Submissions
+    ( Submissions (Submissions), transactionsL )
+import Cardano.Wallet.Submissions.TxStatus
+    ( HasTxId (..), getTx, _Expired, _InLedger, _InSubmission )
+import Control.Arrow
+    ( (&&&) )
+import Control.Lens
+    ( lastOf, (&), _2 )
+import Control.Lens.Extras
+    ( is )
+import System.Random
+    ( Random )
+import Test.QuickCheck
+    ( Arbitrary (arbitrary), Gen, choose, elements, frequency, getSize )
+import Test.QuickCheck.Property
+    ( Property, conjoin, cover, forAllShrinkShow, mapSize )
+
+import Data.Foldable
+    ( toList )
+import qualified Data.Map.Strict as Map
+import Data.Maybe
+    ( fromJust )
+import qualified Data.Text.Lazy as T
+import Text.Pretty.Simple
+    ( pShow )
+
+newtype Slot = Slot Int
+    deriving (Eq, Show, Ord, Random, Enum, Num, Arbitrary)
+
+newtype Tx = Tx Int
+    deriving (Eq, Show, Ord, Arbitrary)
+
+type P x = x Slot Tx
+
+instance HasTxId Tx where
+    type TxId Tx = Tx
+    txId x = x
+
+txG
+    :: Arbitrary tx
+    => Int -- ^ unknown share
+    -> Int -- ^ in submissions share
+    -> Int -- ^ in ledger share
+    -> Int -- ^ expired share
+    -> Submissions slot tx -- ^ source
+    -> Gen tx  -- ^ choosen tx from source
+txG new oldInS oldInL oldE (Submissions db _finality _tip) =
+    frequency $
+        [(new,  arbitrary)]
+        <> include oldInS (is _InSubmission)
+        <> include oldInL (is _InLedger)
+        <> include oldE (is _Expired)
+    where include n l = onNonEmpty n (fmap (fromJust . getTx) $ toList $ Map.filter l db)
+
+onNonEmpty :: Int -> [a] -> [(Int, Gen a)]
+onNonEmpty _ [] = []
+onNonEmpty k xs = [(k, elements xs)]
+
+slotG
+    :: (Random slot, Num slot) => Int -- ^ before the finality share
+    -> Int -- ^ between finality and tip share
+    -> Int -- ^ after the tip share
+    -> Submissions slot tx -- ^ source of tip and finality
+    -> Gen slot -- ^ selected slot
+slotG bf bft at (Submissions _db finality' tip') =
+    frequency
+        [ (bf, choose (finality' - 5, finality'))
+        , (bft, choose (finality' + 1, tip'))
+        , (at, choose (tip' + 1, tip' + 10))
+        ]
+
+
+-- | test parameters
+data SubmissionsHG a = SubmissionsHG
```

Comment:
```suggestion
data GenSubmissionsHistory delta = GenSubmissionsHistory
```

These suggested names would convey more meaning? 🤔 I fully don't understand what the `H` means — probably `History`?